ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/SystematicAnalysisEngine.java

```diff
@@ -141,23 +147,27 @@ private void checkSensiResults(LinearRaoData linearRaoData, SystematicSensitivit
         }
     }
 
-    private Map<String, Double> computeLoopflowAndCheckLoopflowConstraint(LinearRaoData linearRaoData) {
-        Map<String, Double> loopflows;
-        if (!Objects.isNull(linearRaoParameters.getExtendable()) && linearRaoParameters.getExtendable().isLoopflowApproximation()) { //no re-compute ptdf
-            loopflows = new LoopFlowComputation(linearRaoData.getCrac()).calculateLoopFlowsApproximation(linearRaoData.getNetwork());
-        } else {
-            loopflows = new LoopFlowComputation(linearRaoData.getCrac()).calculateLoopFlows(linearRaoData.getNetwork()); //re-compute ptdf
-        }
-
+    /**
+     * check loopflow here with the initial constraint which does not contain loopflowConstraintAdjustmentCoefficient;
+     *  - if LinearRao has run already: MPConstraint for linear rao solver contains loopflowConstraintAdjustmentCoefficient, here is a double-check without loopflowConstraintAdjustmentCoefficient;
+     *  - if LinearRao is skipped, here is the only check for loopflow.
+     */
+    public void checkLoopflowViolationAndSetLoopflowVirtualCost(LinearRaoData linearRaoData, Map<String, Double> loopflows) {
+        setLoopflowViolation(false); //reset loopflow violation status
+        double loopflowTotalViolationCost = 0.0;
+        double loopflowUnitViolationCost = linearRaoParameters.getExtendable().getLoopflowViolationCost();
         for (Cnec cnec : linearRaoData.getCrac().getCnecs(linearRaoData.getCrac().getPreventiveState())) {
             if (!Objects.isNull(cnec.getExtension(CnecLoopFlowExtension.class))
                     && Math.abs(loopflows.get(cnec.getId())) > Math.abs(cnec.getExtension(CnecLoopFlowExtension.class).getLoopFlowConstraint())) {
                 setLoopflowViolation(true);
-                LOGGER.info("Some loopflow constraints are not respected.");
-                break;
+                loopflowTotalViolationCost += loopflowUnitViolationCost * (Math.abs(loopflows.get(cnec.getId())) - Math.abs(cnec.getExtension(CnecLoopFlowExtension.class).getLoopFlowConstraint()));
             }
         }
-        return loopflows;
+
+        linearRaoData.getCracResult().setVirtualCost(loopflowTotalViolationCost); // "zero-loopflowViolationCost", no virtual cost available from Linear optim, set to MAX
+        if (isLoopflowViolation() && linearRaoParameters.getExtendable().getLoopflowViolationCost() == 0.0) {
+            linearRaoData.getCracResult().setVirtualCost(DEFAULT_LOOPFLOWVIOLATION_VIRTUALCOST); // "zero-loopflowViolationCost", no virtual cost available from Linear optim, set to MAX
+        }
```

Comment:
What is that ? And why do you define two different values for default violation cost ?

Reply:
See below